### PR TITLE
Allow conflicting extras in explicit index assignments

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR enables something like the "final boss" of PyTorch setups -- explicit support for CPU vs. GPU-enabled variants via extras:

```toml
[project]
name = "project"
version = "0.1.0"
requires-python = ">=3.13.0"
dependencies = []

[project.optional-dependencies]
cpu = [
    "torch==2.5.1+cpu",
]
gpu = [
    "torch==2.5.1",
]

[tool.uv.sources]
torch = [
    { index = "torch-cpu", extra = "cpu" },
    { index = "torch-gpu", extra = "gpu" },
]

[[tool.uv.index]]
name = "torch-cpu"
url = "https://download.pytorch.org/whl/cpu"
explicit = true

[[tool.uv.index]]
name = "torch-gpu"
url = "https://download.pytorch.org/whl/cu124"
explicit = true

[tool.uv]
conflicts = [
    [
        { extra = "cpu" },
        { extra = "gpu" },
    ],
]
```

It builds atop the conflicting extras work to allow sources to be marked as specific to a dedicated extra being enabled or disabled.

As part of this work, sources now have an `extra` field. If a source has an `extra`, it means that the source is only applied to the requirement when defined within that optional group. For example, `{ index = "torch-cpu", extra = "cpu" }` above only applies to `"torch==2.5.1+cpu"`.

The `extra` field does _not_ mean that the source is "enabled" when the extra is activated. For example, this wouldn't work:

```toml
[project]
name = "project"
version = "0.1.0"
requires-python = ">=3.13.0"
dependencies = ["torch"]

[tool.uv.sources]
torch = [
    { index = "torch-cpu", extra = "cpu" },
    { index = "torch-gpu", extra = "gpu" },
]

[[tool.uv.index]]
name = "torch-cpu"
url = "https://download.pytorch.org/whl/cpu"
explicit = true

[[tool.uv.index]]
name = "torch-gpu"
url = "https://download.pytorch.org/whl/cu124"
explicit = true
```

In this case, the sources would effectively be ignored. Extras are really confusing... but I think this is correct? We don't want enabling or disabling extras to affect resolution information that's _outside_ of the relevant optional group.

### CODE REVIEW

**Comment by charliermarsh** [comment]
This needs documentation, but I'm gonna wait for approval.

**Comment by zanieb** [comment]
I worry about the confusion of

> The extra field does not mean that the source is "enabled" when the extra is activated.

But otherwise I'm into it! We could be more verbose and say `from-extra` to help clarify that case, but I don't know if I love that.

**Comment by charliermarsh** [comment]
This now also supports dependency groups equally.

**Comment by charliermarsh** [comment]
I could also try to make it such that we error if we see the dependency requested as a production dependency, when it has a source with an extra on it?

**Comment by zanieb** [comment]
> from = { extra = "cpu" } and from = { extra = "gpu" }?

I can't think of a case from `from = <package>` makes sense which would be a mismatch with our other uses of `from`. Hm.

> I could also try to make it such that we error if we see the dependency requested as a production dependency, when it has a source with an extra on it?

This seems reasonable, or at least a warning.

**Comment by charliermarsh** [comment]
Ok, I'll start by adding some dedicated errors around this stuff. I think with good error messages (e.g., detect if `torch` is in `project.dependencies` but not in the relevant extra) it could be fine.

**Comment by charliermarsh** [comment]
Ok, I added some dedicated error messages around this.

File: crates/uv-resolver/src/resolver/indexes.rs
Comment (konstin) [inline crates/uv-resolver/src/resolver/indexes.rs:28]: This is the subset of forks in which this entry occurs, not a conflict we generated?

File: crates/uv-pep508/src/marker/tree.rs
Comment (konstin) [inline crates/uv-pep508/src/marker/tree.rs:1134]: Is that reachable?

**Review by konstin — APPROVED** [review]
This is a cool application of conflicting extras!

File: crates/uv-distribution/src/metadata/lowering.rs
Comment (BurntSushi) [inline crates/uv-distribution/src/metadata/lowering.rs:46]: Should these be `Option<&ExtraName>` and `Option<&GroupName>`? It looks like the Clippy lint for this was suppressed explicitly, so I'm guessing there's a reason for it, but it doesn't immediately stand out to me.

File: crates/uv-resolver/src/resolver/indexes.rs
Comment (BurntSushi) [inline crates/uv-resolver/src/resolver/indexes.rs:71]: Nice

**Review by BurntSushi — APPROVED** [review]
Nice! I think this all makes sense to me.

**Comment by zanieb** [comment]
Just to clarify, this looks like it includes `group = ` as well as `extra = ` support?

File: crates/uv-distribution/src/metadata/lowering.rs
Comment (charliermarsh) [reply crates/uv-distribution/src/metadata/lowering.rs:46]: They should but I can't for the life of me figure out how to get it to work at the call site.

Comment (charliermarsh) [reply crates/uv-distribution/src/metadata/lowering.rs]: Figured it out by removing a lifetime.

Comment (BurntSushi) [reply crates/uv-distribution/src/metadata/lowering.rs]: Ahhh yeah I see now. Nice.

**Pull request merged**
